bson/_cbsonmodule.c

```diff
@@ -1659,19 +1659,19 @@ void handle_invalid_doc_error(PyObject* dict) {
     if (evalue && PyErr_GivenExceptionMatches(etype, InvalidDocument)) {
         PyObject *msg = PyObject_Str(evalue);
         if (msg) {
-            // Add doc to the error class as a property.
-            PyObject_SetAttrString(InvalidDocument, "document", dict);
             const char * msg_utf8 = PyUnicode_AsUTF8(msg);
             if (msg_utf8 == NULL) {
                 goto cleanup;
             }
             PyObject *new_msg = PyUnicode_FromFormat("Invalid document: %s", msg_utf8);
+            // Add doc to the error instance as a property.
+            PyObject *new_evalue = PyObject_CallFunctionObjArgs(InvalidDocument, new_msg, dict, NULL);
```

Comment:
Need to check the `new_msg == NULL` case before we call PyObject_CallFunctionObjArgs here.